src/main/java/com/gluonhq/substrate/util/FileOps.java

```diff
@@ -365,4 +373,127 @@ public static void extractFilesFromJar(String extension, Path sourceJar, Path ta
+                }
+                zipEntry = zis.getNextEntry();
+            }
+            zis.closeEntry();
+        }
+        return hashes;
+    }
+
+    /**
+     * Downloads a zip file from the specified zipUrl into the targetPath where a file
+     * named zipName will be created.
+     * Note the conventions for the url file (must end with .zip) and the destination file
+     * (must end with .zip)
+     * Once the zip file is downloaded, it will be unpacked into the location that starts at
+     * the targetPath, and is resolved under targetPath/level0/level1/level2
+     * where ether level1 or level2 can be null.
+     * A file with the checksums of all the files in the zip will be generated with extension
+     * ".md5" under the final path: targetPath/level0/level1/level2/level0-level2.md5
+     *
+     * @param zipUrl a string with the location of a zip file, e.g. https://download2.gluonhq.com/substrate/bar/foo.zip
+     * @param targetPath the path where the file zip file will be downloaded, e.g. /opt/bar
+     * @param zipName the name of the file that will be downloaded, e.g. foo.zip
+     * @param level0 the folder under targetPath, not null, e.g. foo1
+     * @param level1 a folder under level0, it can be null, e.g. foo2
+     * @param level2 a folder under level1, it can be null, e.g. foo3
+     *              The zip file will be downloaded into /opt/bar/foo.zip
+     *              The contents of this zip file will be installed into
+     *              /opt/bar/foo1/foo2/foo3, and also the file
+     *               /opt/bar/foo1/foo2/foo3/foo1-foo3.md5 will be created
+     * @throws IOException
+     */
+    public static void processZip(String zipUrl, Path targetPath, String zipName,
+                                  String level0, String level1, String level2) throws IOException {
+        Objects.requireNonNull(level0);
+
+        String md5name = level2 == null ? level0 + ".md5" : level0 + "-" + level2 + ".md5";
+        Path zipPath = targetPath.resolve(zipName);
+        Logger.logDebug("Processing zip file: url = " + zipUrl +
+                ", zip = " + zipPath +
+                ", level0 = " + level0 +
+                ", level1 = " + level1 +
+                ", level2 = " + level2 +
+                ", md5 = " + md5name);
+
+        // 1. Download zip from urlZip into zipPath
+        FileOps.downloadFile(new URL(zipUrl), zipPath);
+
+        // 2. Set path where zip should be extracted
+        Path zipDir = targetPath.resolve(level0);
+        if (level1 != null) zipDir = zipDir.resolve(level1);
+        if (level2 != null) zipDir = zipDir.resolve(level2);
+
+        Files.createDirectories(zipDir);
+
+        // 3. Extract zip from zipPath into zipDir
+        Map<String, String> hashes = FileOps.unzipFile(zipPath, zipDir);
+
+        // 4. Write hashes file into zipDir
+        try (FileOutputStream fos =
+                     new FileOutputStream(zipDir.toString() + File.separator + md5name);
```

Comment:
Much safer to use `Path` API for the name to avoid using `Path.separator` IMO

Reply:
Ok, done